### PR TITLE
Attack system

### PR DESCRIPTION
**Pull request by SamuelllG**
- implemented reusable attack system for player and enemies
- adapted playercontroller and enemycontroller to use attacklogic
- implemented IDamageable and IMovable interface (IMovable could also be an abstract class and inherit 4-5 methods which are the same for player and enemy)
- implemented hit marker / animation replacement as material (color) change (this is in Stats class which will be replaced with real stats class)

### CODE REVIEW

File: Assets/Scripts/AttackLogic.cs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    // ----temporary as animation replacement------
+        private MeshRenderer _gameObjectRenderer;
+        private Material _prevMat;
+        private Material _attackAnimationMaterial;
+    // ---------
+
+    private void Awake()
+    {
+        // init components
+        _movable = GetComponent<IMovable>();
```
Comment (DavidPgl) [inline Assets/Scripts/AttackLogic.cs:51]: Always a good idea to add `[RequiresComponent(typeOf(IMovable))` to the class, if you use GetComponent in the Awake

File: Assets/Scripts/AttackLogic.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+        // -----temp replacement for animation-----
+            _gameObjectRenderer = gameObject.GetComponentInChildren<MeshRenderer>();
```
Comment (DavidPgl) [inline Assets/Scripts/AttackLogic.cs:54]: Please use a reference:(

File: Assets/Scripts/AttackLogic.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        else if (PerformingHit)
+        {
+            bool? success = PerformHit();
+            if (success != null)
+            {
+                // hit animation performed
+                if (success == true)
+                {
+                    // deal damage
+                    IDamageable damageable = (IDamageable)_target.GetComponent<IDamageable>();
```
Comment (DavidPgl) [inline Assets/Scripts/AttackLogic.cs]: You could define _target as IDamageable and save one GetComponent call. GetComponent is very heavy. If you can avoid it, try to.

File: Assets/Scripts/AttackLogic.cs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    public void StopAttack()
+    {
+        _target = null;
+    }
+
+    /// <summary>
+    /// Waits (attackTime) for hit to be finished and returns the result.
+    /// </summary>
+    /// <returns>Null if hit not finished, true if hit successful performed, false if unsuccessful</returns>
+    private bool? PerformHit()
```
Comment (DavidPgl) [inline Assets/Scripts/AttackLogic.cs]: Probably personal preference but not the biggest fan of making bool to a three-state type. This is probably one of the few cases where using it is still readable. Defining three Enums for this would probably look nicer though:
`AttackStatus.Hit`
`AttackStatus.Miss`
`AttackStatus.NotFinished`
or something like that.
Also preferences but Unity offers **Coroutines**, in which you can wait for a time or a specified event to be completed. But I'm not sure if you can use it here

File: Assets/Scripts/AttackLogic.cs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    {
+        _timer += Time.deltaTime;
+
+        if (_timer > attackTime)
+        {
+            // animation finished
+            _timer = 0;
+            // target still in range?
+            float distanceToTarget = Vector3.Distance(_target.transform.position, transform.position);
```
Comment (DavidPgl) [inline Assets/Scripts/AttackLogic.cs]: You calculate the distance each frame anywhere (line 78), so saving it in a variable would save you one calculation

File: Assets/Scripts/Interfaces/IDamageable.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Interfaces
+{
+    public interface IDamageable
```
Comment (DavidPgl) [inline Assets/Scripts/Interfaces/IDamageable.cs]: Would an abstract class be useful here? I mean enemies and the player both share health and the "loosing health behavior".

File: Assets/Scripts/Interfaces/IMovable.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Interfaces
+{
+    public interface IMovable
```
Comment (DavidPgl) [inline Assets/Scripts/Interfaces/IMovable.cs:7]: Same as above. Only if the methods do the same thing of course.

File: Assets/Scripts/Player/PlayerController.cs [hunk 8]
```diff
@@ -83,30 +102,106 @@ private void Update()
...
+            if (Physics.Raycast(clickRay, out RaycastHit hit, 10000) && !_attackLogic.PerformingHit)
             {
-                // Set destination for nav mesh agent
-                _navMeshAgent.SetDestination(hit.point);
-                _navMeshAgent.isStopped = false;
-
-                // Create click point effect
-                Instantiate(clickEffect, hit.point + Vector3.up * 5, Quaternion.identity);
+                GameObject objectHit = hit.collider.gameObject;
+                IDamageable damageable = (IDamageable)objectHit.GetComponent<IDamageable>();
```
Comment (DavidPgl) [inline Assets/Scripts/Player/PlayerController.cs]: Is the cast really necessary?

File: Assets/Scripts/Player/PlayerController.cs [hunk 9]
```diff
@@ -83,30 +102,106 @@ private void Update()
...
+        /// <param name="target">The target to face</param>
+        /// <param name="shouldTurn">Whether the object should turn to the target or not</param>
+        /// <returns>Whether the object is facing the target</returns>
+        public bool FaceTarget(GameObject target, bool shouldTurn)
+        {
+            float f;
+            return FaceTarget(target, shouldTurn, out f);
+        }
+
+        // ----- Note: same in EnemyController --> make IMovable abstract class and inherit? ------
```
Comment (DavidPgl) [inline Assets/Scripts/Player/PlayerController.cs:180]: Sounds good to me

File: Assets/Scripts/Player/PlayerController.cs [hunk 10]
```diff
@@ -83,30 +102,106 @@ private void Update()
...
+        /// <summary>
+        /// Faces the target. Returns true if facing the target.
+        /// </summary>
+        /// <param name="target">The target to face</param>
+        /// <param name="shouldTurn">Whether the object should turn to the target or not</param>
+        /// <returns>Whether the object is facing the target</returns>
+        public bool FaceTarget(GameObject target, bool shouldTurn)
+        {
+            float f;
+            return FaceTarget(target, shouldTurn, out f);
```
Comment (DavidPgl) [inline Assets/Scripts/Player/PlayerController.cs]:
```suggestion
            return FaceTarget(target, shouldTurn, out _);
```
Then you can delete the local variable

File: Assets/Scripts/Player/Stats.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
+﻿using Interfaces;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
```
Comment (DavidPgl) [inline Assets/Scripts/Player/Stats.cs]: Namespace missing

File: Assets/Scripts/Player/Stats.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        // just put initial mat here
+        _prevMat = _gameObjectRenderer.material;
+    }
+
+    /// <summary>
+    /// Changes the objects color back to normal after being hit.
+    /// </summary>
+    private void Update()
+    {
+        if (_hit)
```
Comment (DavidPgl) [inline Assets/Scripts/Player/Stats.cs]:
```suggestion
        if (!_hit) return
```
This way it will skip the rest

File: Assets/Scripts/AttackLogic.cs [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+            // chase target
+            _movable.Move(_target.transform.position);
+        }
+    }
+
+    /// <summary>
+    /// Starts the attack.
+    /// </summary>
+    /// <param name="target">The target to be attacked</param>
+    public void StartAttack(GameObject target)
```
Comment (DavidPgl) [inline Assets/Scripts/AttackLogic.cs:157]:
```suggestion
    public void StartAttack(IDamageable target)
```
Make it as specific as possible

**Review by DavidPgl — CHANGES_REQUESTED** [review]

**Review by thuemmels — DISMISSED** [review]

File: Assets/Scripts/AttackLogic.cs (hunk 2 above)
Comment (SamuelllG) [reply Assets/Scripts/AttackLogic.cs:54]: over inspector? but it's just temporary...

File: Assets/Scripts/AttackLogic.cs (hunk 3 above)
Comment (SamuelllG) [reply Assets/Scripts/AttackLogic.cs]: How do I save one GetComponent call then? the only point where I call GetComponent is when the hit was performed successfully, if I define _target as IDamageable, I'd have to GetComponent everytime independent of the hit being successful or not. Also this would cause trouble at different places like calculating the distance etc.

File: Assets/Scripts/AttackLogic.cs (hunk 13 above)
Comment (SamuelllG) [reply Assets/Scripts/AttackLogic.cs:157]: as stated above, same thing

File: Assets/Scripts/AttackLogic.cs (hunk 4 above)
Comment (SamuelllG) [reply Assets/Scripts/AttackLogic.cs]: Done, but doesn't improve readability in my opinion. Also removed PerformingHit variable and put it all in one: Status.

File: Assets/Scripts/AttackLogic.cs (hunk 5 above)
Comment (SamuelllG) [reply Assets/Scripts/AttackLogic.cs]: Indeed, however I always try to keep the number of instance variables low (matter of readability), and a parameter doesn't really make sense. But I guess, if it's better to make one, I'll make one ^^

File: Assets/Scripts/Interfaces/IDamageable.cs (hunk 6 above)
Comment (SamuelllG) [reply Assets/Scripts/Interfaces/IDamageable.cs]: I thought about that as well, maybe we have to discuss this. In my opinion not a matter of this PR, we can create a separate issue for this :)

File: Assets/Scripts/Interfaces/IMovable.cs (hunk 7 above)
Comment (SamuelllG) [reply Assets/Scripts/Interfaces/IMovable.cs:7]: yep

File: Assets/Scripts/Player/PlayerController.cs (hunk 8 above)
Comment (SamuelllG) [reply Assets/Scripts/Player/PlayerController.cs]: nope xD

File: Assets/Scripts/Player/PlayerController.cs (hunk 9 above)
Comment (SamuelllG) [reply Assets/Scripts/Player/PlayerController.cs:180]: we should a separate issue for this

File: Assets/Scripts/Player/Stats.cs (hunk 11 above)
Comment (SamuelllG) [reply Assets/Scripts/Player/Stats.cs]: yep, I will integrate this into PlayerState

File: Assets/Scripts/Player/Stats.cs (hunk 12 above)
Comment (SamuelllG) [reply Assets/Scripts/Player/Stats.cs]: and the advantage is? skip one if statement? I rather take the readability, thanks 😄 (I changed it a bit)

**Comment by SamuelllG** [comment]
I also merged master which was a bit of a pain 😅 and fixed everything. I added the abstract class discussion as a task as well as implementing trigger colliders for the NPCs :)

File: Assets/Scripts/AttackLogic.cs (hunk 2 above)
Comment (Kaes3kuch3n) [reply Assets/Scripts/AttackLogic.cs:54]: For temporary stuff I'd suggest adding a //TODO or //BUG so that we can find stuff like this easily later

File: Assets/Scripts/AttackLogic.cs [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+    /// <summary>
+    /// Ends the hit. Deals damage to the target, if hit was successful. Resets aggro if necessary.
+    /// </summary>
+    private void HitFinished()
+    {
+        // hit animation performed
+        if (Status == AttackStatus.Hit)
+        {
+            // deal damage
+            IDamageable damageable = (IDamageable)_target.GetComponent<IDamageable>();
```
Comment (Kaes3kuch3n) [inline Assets/Scripts/AttackLogic.cs]: I'm pretty sure this cast ist redundant, since GetComponent<IDamageable> returns an object of type IDamageable 🤔

File: Assets/Scripts/AttackLogic.cs (hunk 13 above)
Comment (Kaes3kuch3n) [reply Assets/Scripts/AttackLogic.cs:157]: Could you please elaborate on that? It's not clear at all to me what you mean by that/what you're referring to 😅 

Reply:
You sound like Prof. WeWu :D
Since I couldn't reply, here with an edit: I added a few words to the summary. "If the target is not in attacking range, chases it. If the target is near performs a hit on it and either resets or continues attacking." -- just the normal "attack cycle".

File: Assets/Scripts/EnemyController.cs [hunk 15]
```diff
@@ -150,27 +164,65 @@ private void MoveToNextPoint()
...
+    /// <summary>
+    /// Faces the target. Returns true if facing the target.
+    /// </summary>
+    /// <param name="target">The target to face</param>
+    /// <param name="shouldTurn">Whether the object should turn to the target or not</param>
+    /// <returns>Whether the object is facing the target</returns>
+    public bool FaceTarget(GameObject target, bool shouldTurn)
+    {
+        float f;
+        return FaceTarget(target, shouldTurn, out f);
```
Comment (Kaes3kuch3n) [inline Assets/Scripts/EnemyController.cs]: If you don't wanna use an out variable, you can simply use `out _` to discard the result

File: Assets/Scripts/Interfaces/IMovable.cs [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        void Move(Vector3 position);
+
+        bool FaceTarget(GameObject target, bool shouldTurn, out float difference);
+
+        bool FaceTarget(GameObject target, bool shouldTurn);
```
Comment (Kaes3kuch3n) [inline Assets/Scripts/Interfaces/IMovable.cs]: Feels kinda redundant with the other FaceTarget method defined right above it, since we can just discard the out value if we don't need it

File: Assets/Scripts/Player/PlayerController.cs [hunk 17]
```diff
@@ -114,26 +132,101 @@ private void SetUpControls()
...
         /// </summary>
-        private void Move(InputAction.CallbackContext obj)
+        private void OnRightClick(InputAction.CallbackContext obj)
         {
             Ray clickRay = _camera.ScreenPointToRay(Mouse.current.position.ReadValue());
 
-            // Get ground position from mouse click
-            if (Physics.Raycast(clickRay, out RaycastHit hit, 10000, moveClickLayers))
+            // only change target / move, if not performing a hit
+            if (Physics.Raycast(clickRay, out RaycastHit hit, 10000) && _attackLogic.Status == AttackLogic.AttackStatus.None)
```
Comment (Kaes3kuch3n) [inline Assets/Scripts/Player/PlayerController.cs:145]: Would maybe change this in the future to make it possible to cancel attacks by moving away from the enemy. Doesn't really feel like the player is in control like this 😬

File: Assets/Scripts/Player/PlayerController.cs [hunk 18]
```diff
@@ -154,11 +247,11 @@ private void RotateCamera(InputAction.CallbackContext obj)
...
         private void UpdateCameraAngle(float cameraDistanceChange = 0)
         {
-            float radian = (float) Math.PI * _cameraAngleX / 180;
+            float radian = (float)Math.PI * _cameraAngleX / 180;
             cameraDistance.y = Mathf.Clamp(cameraDistance.y + cameraDistanceChange, cameraDistanceRange.min,
                 cameraDistanceRange.max);
-            _cameraPosition = new Vector3((float) Math.Sin(radian) * cameraDistance.x, cameraDistance.y,
-                (float) -Math.Cos(radian) * cameraDistance.x);
+            _cameraPosition = new Vector3((float)Math.Sin(radian) * cameraDistance.x, cameraDistance.y,
+                (float)-Math.Cos(radian) * cameraDistance.x);
```
Comment (Kaes3kuch3n) [inline Assets/Scripts/Player/PlayerController.cs:243]: Why did you remove the spaces? 😂

File: Assets/Scripts/PlayerManager.cs [hunk 19]
```diff
@@ -0,0 +1,34 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class PlayerManager : MonoBehaviour
+{
+    #region Singleton
```
Comment (Kaes3kuch3n) [inline Assets/Scripts/PlayerManager.cs]: Could you switch to David's Singleton base class for this logic?

**Review by Kaes3kuch3n — CHANGES_REQUESTED** [review]

File: Assets/Scripts/AttackLogic.cs (hunk 2 above)
Comment (DavidPgl) [reply Assets/Scripts/AttackLogic.cs:54]: It only takes a minute. Don't be lazy:P

File: Assets/Scripts/AttackLogic.cs (hunk 14 above)
Comment (SamuelllG) [reply Assets/Scripts/AttackLogic.cs]: for sure, ye I forgot that here

File: Assets/Scripts/EnemyController.cs (hunk 15 above)
Comment (SamuelllG) [reply Assets/Scripts/EnemyController.cs]: forgot that here as well

File: Assets/Scripts/Interfaces/IMovable.cs (hunk 16 above)
Comment (SamuelllG) [reply Assets/Scripts/Interfaces/IMovable.cs]: good point

File: Assets/Scripts/Player/PlayerController.cs (hunk 17 above)
Comment (SamuelllG) [reply Assets/Scripts/Player/PlayerController.cs:145]: We can discuss about that. It might be useful, idk. It's a tiny change howsoever.

File: Assets/Scripts/Player/PlayerController.cs (hunk 18 above)
Comment (SamuelllG) [reply Assets/Scripts/Player/PlayerController.cs:243]: I didn't, must have come from merging.

File: Assets/Scripts/PlayerManager.cs (hunk 19 above)
Comment (SamuelllG) [reply Assets/Scripts/PlayerManager.cs]: didn't see there is already one - switched

**Review by Kaes3kuch3n — DISMISSED** [review]

File: Assets/Scripts/AttackLogic.cs (hunk 3 above)
Comment (DavidPgl) [reply Assets/Scripts/AttackLogic.cs]: _target is of type GameObject and will be set at some point. Is there a reason why you can't define it as a IDamageable from the beginning? Attacking will happen more often than setting the _target, right?

File: Assets/Scripts/EnemyController.cs [hunk 20]
```diff
@@ -47,9 +52,13 @@ public class EnemyController : MonoBehaviour
     [SerializeField]
     private bool showEscapeRadius;
 
-    private Transform _target;
+    // component references
     private NavMeshAgent _agent;
     private WanderAI _wanderAI;
+    private AttackLogic _attackLogic;
```
Comment (DavidPgl) [inline Assets/Scripts/EnemyController.cs:58]: So this script accesses AttackLogic and AttackLogic accesses the EnemyController (with IMovable). Does the AttackLogic really care about IMovable? Wouldn't it be enough if the EnemyController activates the attacking and stops moving?

File: Assets/Scripts/Managers/PlayerManager.cs [hunk 21]
```diff
@@ -1,21 +1,16 @@
...
-using UnityEngine;
-using UnityEngine.Networking.PlayerConnection;
+﻿using UnityEngine;
 
 namespace Managers
 {
     public class PlayerManager : Singleton<PlayerManager>
     {
-        [SerializeField] private GameObject player;
+        [SerializeField] private GameObject _player;
```
Comment (DavidPgl) [inline Assets/Scripts/Managers/PlayerManager.cs]: SerializedFields are normaly without underscore. Was already correct:)

File: Assets/Scripts/Player/EnemyState.cs [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+/// </summary>
+public class EnemyState : MonoBehaviour, IDamageable
+{
+    [Tooltip("The time the object should be marked as hit after being hit")]
+    [SerializeField]
+    private float hitMarkTime;
+
+    [Tooltip("The MeshRenderer of the graphics object of the player")]
+    [SerializeField]
+    private MeshRenderer _gameObjectRenderer;
```
Comment (DavidPgl) [inline Assets/Scripts/Player/EnemyState.cs]: Stick with SerializedField without underscore

File: Assets/Scripts/Player/PlayerState.cs [hunk 23]
```diff
@@ -21,7 +21,24 @@ public class PlayerState : MonoBehaviour, IDamageable
...
+
+
+        // ------temp for hit animation------
+        [Tooltip("The time the object should be marked as hit after being hit")]
+        [SerializeField]
+        private float hitMarkTime;
+
+        [Tooltip("The MeshRenderer of the graphics object of the player")]
+        [SerializeField]
+        private MeshRenderer _gameObjectRenderer;
```
Comment (DavidPgl) [inline Assets/Scripts/Player/PlayerState.cs]:
```suggestion
        private MeshRenderer gameObjectRenderer;
```

File: Assets/Scripts/Player/EnemyState.cs (hunk 22 above)
Comment (DavidPgl) [reply Assets/Scripts/Player/EnemyState.cs]:
```suggestion
    private MeshRenderer gameObjectRenderer;
```

File: Assets/Scripts/Managers/PlayerManager.cs (hunk 21 above)
Comment (DavidPgl) [reply Assets/Scripts/Managers/PlayerManager.cs]:
```suggestion
        [SerializeField] private GameObject player;
```

**Comment by DavidPgl** [comment]
Das ist das letzte :P

File: Assets/Scripts/Managers/PlayerManager.cs (hunk 21 above)
Comment (SamuelllG) [reply Assets/Scripts/Managers/PlayerManager.cs]: oh ye sure :joy: my bad

File: Assets/Scripts/Player/EnemyState.cs (hunk 22 above)
Comment (SamuelllG) [reply Assets/Scripts/Player/EnemyState.cs]: dang, what did I do

File: Assets/Scripts/Managers/PlayerManager.cs (hunk 21 above)
Comment (SamuelllG) [reply Assets/Scripts/Managers/PlayerManager.cs]: for a singleton I don't expect a serialize

File: Assets/Scripts/EnemyController.cs (hunk 20 above)
Comment (SamuelllG) [reply Assets/Scripts/EnemyController.cs:58]: That's a good point, but AttackLogic needs IMovable anyway, so that it has a target which it can turn too and which it can check the position after completing the hit :)

I look at it like this: 
You click on the ground --> you move --> Controller
You click on an enemy --> you attack --> AttackLogic
If you have to move in order to attack the enemy that's part of the AttackLogic. Otherwise you'd have to implement something like "target out of range" when clicking on an enemy far away.

Comment (DavidPgl) [reply Assets/Scripts/EnemyController.cs:58]: Ahh I see, I was always thinking this way:
You click on something: ground or enemy or resource (Not in range) -->Move to it --> Movable
You click on something (In range) --> Trigger action (e.g. attack or collect) --> specific script/component
Both works I guess

**Review by DavidPgl — APPROVED** [review]

File: Assets/Scripts/AttackLogic.cs (hunk 2 above)
Comment (SamuelllG) [reply Assets/Scripts/AttackLogic.cs:54]: haha you're talking about two different things ^^
I made it serializable, Luis talks about the whole animation / temporary color thingy

File: Assets/Scripts/AttackLogic.cs (hunk 3 above)
Comment (SamuelllG) [reply Assets/Scripts/AttackLogic.cs]: I don't know if setting the target (clicking on the enemy) will happen less often than a successful hit... especially because I know people that rape the mouse button on attack :D

(I need the gameobject so that I have the position of the enemy and that I can calculate the distance.)

File: Assets/Scripts/AttackLogic.cs (hunk 13 above)
Comment (SamuelllG) [reply Assets/Scripts/AttackLogic.cs:157]: Oh okay it also didn't show me that you commented on existing comments... what I'm talking about is the discussion of whether to use IDamageable as target or GameObject (see the only open discussion above)

**Review by Kaes3kuch3n — APPROVED** [review]
